src/modules/mavlink/mavlink_receiver.cpp

```diff
@@ -1624,34 +1624,52 @@ MavlinkReceiver::handle_message_hil_state_quaternion(mavlink_message_t *msg)
 void *
 MavlinkReceiver::receive_thread(void *arg)
 {
-#ifndef __PX4_POSIX
-	int uart_fd = _mavlink->get_uart_fd();
 
 	const int timeout = 500;
 	uint8_t buf[32];
-
 	mavlink_message_t msg;
 
-	/* set thread name */
-	char thread_name[24];
-	sprintf(thread_name, "mavlink_rcv_if%d", _mavlink->get_instance_id());
-	prctl(PR_SET_NAME, thread_name, getpid());
-
 	struct pollfd fds[1];
-	fds[0].fd = uart_fd;
-	fds[0].events = POLLIN;
+
+	struct sockaddr_in srcaddr;
+	socklen_t addrlen = sizeof(srcaddr);
+
+	int uart_fd;
+
+	if (_mavlink->get_protocol() == SERIAL) {
+		uart_fd = _mavlink->get_uart_fd();
+#ifndef __PX4_POSIX
+		/* set thread name */
+		char thread_name[24];
+		sprintf(thread_name, "mavlink_rcv_if%d", _mavlink->get_instance_id());
+		prctl(PR_SET_NAME, thread_name, getpid());
+#endif
+
+		fds[0].fd = uart_fd;
+		fds[0].events = POLLIN;
+	} else if (_mavlink->get_protocol() == UDP || _mavlink->get_protocol() == TCP) {
+		fds[0].fd = _mavlink->get_socket_fd();
+		fds[0].events = POLLIN;
+	}
 
 	ssize_t nread = 0;
 
 	while (!_mavlink->_task_should_exit) {
-		if (poll(fds, 1, timeout) > 0) {
+		if (poll(&fds[0], 1, timeout) > 0) {
+			if (_mavlink->get_protocol() == SERIAL) {
+				/* non-blocking read. read may return negative values */
+				if ((nread = ::read(uart_fd, buf, sizeof(buf))) < (ssize_t)sizeof(buf)) {
+					/* to avoid reading very small chunks wait for data before reading */
+					usleep(1000);
+				}
+			} else if (_mavlink->get_protocol() == UDP) {
 
-			/* non-blocking read. read may return negative values */
-			if ((nread = ::read(uart_fd, buf, sizeof(buf))) < (ssize_t)sizeof(buf)) {
-				/* to avoid reading very small chunks wait for data before reading */
-				usleep(1000);
+				if (fds[0].revents & POLLIN) {
+					nread = recvfrom(_mavlink->get_socket_fd(), buf, sizeof(buf), 0, (struct sockaddr *)&srcaddr, &addrlen);
```

Comment:
QuRT does not support a network interface in this context. You are right.
QuRT should not build mavlink.
On Jun 19, 2015 7:15 AM, "Lorenz Meier" <notifications@github.com> wrote:

> In src/modules/mavlink/mavlink_receiver.cpp
> <https://github.com/PX4/Firmware/pull/2425#discussion_r32832399>:
>
> >
> > -			/* non-blocking read. read may return negative values */
> > -			if ((nread = ::read(uart_fd, buf, sizeof(buf))) < (ssize_t)sizeof(buf)) {
> > -				/* to avoid reading very small chunks wait for data before reading */
> > -				usleep(1000);
> > +				if (fds[0].revents & POLLIN) {
> > +					nread = recvfrom(_mavlink->get_socket_fd(), buf, sizeof(buf), 0, (struct sockaddr *)&srcaddr, &addrlen);
>
> And here
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/PX4/Firmware/pull/2425/files#r32832399>.
>